### ISSUE TITLE
[WSL] Very frequent "hang" when running any command related to calva

### ISSUE THREAD

**Issue by nathangerday** [BetterThanTomorrow/calva#2331]
I noticed a lot of hangs happening, mostly when navigating through clojure code but all commands related to Calva seem to be concerned. 
While the hang happen I can't do any action related to calva (structural editing or else )  but any other action (basic vscode action or another extension) works fine. Once the hangs finishes, usually after 5 to 10 seconds, all "queued" events will get executed at once.
I can almost consistently reproduce this in a few seconds by moving forward and backward through a forms.

No logs are shown in the console or the extension host output (native and remote).
This is happening in a fairly large project which I can't share unfortunately, I will try to come up with an exemple as simple as possible to see if others can reproduce.
I deactivated all my extensions except Calva and WSL and still run into this issue.

To me, this seems linked to WSL as I have another computer with a very similar setup running the same project but on a native Linux and I never noticed any hang.

As @seancorfield suggested, I also tried running VSCode directly on WSL without remoting into it but visualizing via the X emulation and there doesn't seem to be any hangs that way. ( This is the exact same virtual machine as the one I'm remoting into so the only difference should be the fact that I'm remoting itself if that makes sense)
I will try this more extensively during the next few days.

**Comment by PEZ** [comment]
Thanks for reporting! 🙏

**Comment by nathangerday** [comment]
Hey, sorry for the lack of update. 

I tried to investigate this further but I'm not finding anything conclusive.

I haven't managed to reproduce this issue on any other project than the one I'm working one ( which I unfortunately cannot share ). It has a pretty large size / number of files loaded, so maybe there's some kind of memory limit for the process vscode is hooked to that's only being hit when remoting.

If anyone knows any large opensource Clojure project I could try to reproduce this on, I'd gladly be willing to test it

**Comment by PEZ** [comment]
Hi!

Penpot is pretty large: https://github.com/penpot/penpot

So is Status: https://github.com/status-im/status-mobile

**Comment by PEZ** [comment]
There's also Metabase: https://github.com/metabase/metabase

**Comment by nathangerday** [comment]
I tried all three of the other projects you mentioned. The hang-ups do happen, but they are very insignificant—100ms or 200ms at most—compared to the 2-3 seconds I'm experiencing. That's surprising, considering our project size is very similar to Metabase, and I don't know of anything particular that could cause such a massive difference.

In the last few months, I've tried adjusting most of the settings in Calva, reinstalling it on a different computer with the same WSL configuration, and even using another Linux distro. However, every time I work on this project using remote VSCode with Calva, the issue remains. I'm completely out of ideas.

I’ve included a GIF of the situation in case it helps.
![ezgif-3-705ad358bd](https://github.com/user-attachments/assets/eb10f14b-0939-42ce-a5a1-f28fd5cf4d7f)

I guess the next step would be to run calva in development mode and figure out exactly where it's hanging, I'll probably try this as soon as I have some time.

**Comment by PEZ** [comment]
Does it happen in some particular files? Calva's structural editor performance suffers when there are many forms, especially when they are on the same level. So e.g. editing clojure.core typically is a bit slow. There aren't any 2-3 seconds hangs to my experience, but maybe remote use changes that.

There could also be some particular content. Calva's lexer is regex based and we have had it go into a spin on particular text in the past. All known such cases are fixed, but maybe your code there has found a new case.

Running Calva in development mode sounds like an excellent idea. Let me know if you need any help with that. The instructions in the wiki _should_ be fine, but anyway.

**Comment by nathangerday** [comment]
Thanks !
No, it happens project wide even in some extremely small files such as this : 

```Clojure
(ns electre.diffusion.backend.meta)

(def module-name "diffusion-backend")
(def path "diffusion/backend/")
(def version "0.224.6")
(def tag "diffusion-backend-0.224.6")
(def generated-at "2024-09-24T13:12:35Z")
```

**Comment by PEZ** [comment]
That wins the weirdness-of-the-year prize in my book! 🤔

**Comment by nathangerday** [comment]
Same 😅

Some interesting info: Just tried without connecting to a repl and it doesn't seem to happen at all. I can navigate perfectly fine and absolutly no hangs. ( Though it's not really useful since I mostly can't do anything this way ).
But that does seem to mean that something about the connection is causing the hang regularly and the fact that I'm seeing it when navigating is just based on "luck" depending when it happens ?

**Comment by PEZ** [comment]
Yeah, I don't think Calva is generally waiting for something from the REPL to allow editing. But there's only one thread and it could be stuck on something with the connection. Are you using Jack-in, or are you starting the REPL some other way? That shouldn't matter, but who knows.

Assuming it doesn't matter. Calva opens a connection to the nrepl server and sends and receives messages, converting to bencode before sending and from bencode for incoming messages. And, especially while hovering, Calva talks to the nrepl server to get info about the symbols. So the connection could hang somehow and that could cause Calva to freeze, I guess. Then Calva things should lock up, while (I think) general VS Code things should continue to work. Calva registers an on-type edit provider, so typing gets to be a Calva thing. Something like bringing up the VS Code command palette should still work during the hang. But bringing up Calva's REPL menu (e.g. by clicking the REPL button in the status bar) is a Calva thing.

There's a command for making Calva write out a log of nrepl messages sent and received. *Calva: Toggle nREPL logging* (or something like that). See if that log can give some more info.

**Comment by nathangerday** [comment]
I mostly use Jack-in but did try starting the repl on its own and connecting to it later, and it definitely didn't matter.

Thanks for the suggestion and the explanations. I did try toggling the nREPL logs but didn't see anything in there above a few ms, so it's probably not that.

I tried playing with development mode a bit. First of all, when the hang happens, it's definitely before reaching any calva command, and once reached after a few seconds, the commands run perfectly fine in 2 or 3 ms.

I also tried profiling (I'm pretty new to both typescript and VSCode extension development, so not sure if that's a good way to find this kind of issue here). From what I'm seeing in the results, nothing on the Calva side seems to be wrong, and most of the long times are spent in the extensionHostProcess so this might not be at all relating to anything Calva is doing but just some other issue that is only seen here.

In case it can help, here is the profiling result. During those 40 seconds, I had 2 hangs, both felt about one second long.
[vscode-profile-2024-10-04-20-19-21.cpuprofile](https://github.com/user-attachments/files/17262564/vscode-profile-2024-10-04-20-19-21.cpuprofile)

Sorry for taking up your time if this isn't related to anything you can do.

**Comment by PEZ** [comment]
Thanks for sharing your findings! This is great for me and others to know about, even if there wouldn't be anything Calva can do about it.

I'm wondering if it could be some other extension acting up.

**Comment by nathangerday** [comment]
I have an even weirder conclusion for this issue. 

I just solved it yesterday on 2 different computer though I don't really understand how.
My fix was to install a new Ubuntu 24.01 distro in WSL alongside the 22.04 I was using. Configuring the project in this new version, the hang never happened once. So I figured it might be an issue very specific to Ubuntu 22.

Now here is the weird part. When I went back to the old Ubuntu 22.04 WSL and started the project, I didn't encounter any hang either. 
And to make sure, I reproduced the same steps on my other computer with a similar setup while making sure not to do anything that could impact the old WSL instance. Again, the problem was fixed.

As far as I know, the main system which is not isolated in the WSL VM is docker which runs on the host machine. So maybe something in the configuration got fixed while setting up the new VM ? I really can't figure it out but I'm happy it's fixed.

I think we can close this issue if that's okay with you, If I ever find a definitive answer, I'll make sure to update it. And if anyone else experiences the same bug please don't hesitate to reach out if I may help.

**Comment by PEZ** [comment]
Thanks again! I am totally fine with closing the issue. People hopefully will be able to dig it up through searching.
